Add performance regression test

pgrete opened the pull request:
## PR Summary

The second CI environment at IAS is finally working and this PR adds the basis of performance regression tests automatically creating plots like https://gitlab.com/theias/hpc/jmstone/athena-parthenon/parthenon-ci-mirror/-/jobs/639901312/artifacts/file/build-cuda/tst/regression/outputs/performance/performance.png
![image](https://user-images.githubusercontent.com/22685568/87551864-39d51b80-c6b1-11ea-8764-b9dee0ca8df0.png)

I expect this to expand in the future (in other PRs such as #190 ) so this is just mean to add the infrastructure to the codebase.
For now I left the performance tests as part of the "regression" test as they reuse the python based testing infrastructure.
We could update/refactor is based on #187 as well as while addressing #209 (but I see it as a separate effort from this).

JoshuaSBrown commented on tst/regression/utils/test_case.py:
```diff
@@ -207,20 +208,21 @@ def Run(self):
         sys.stdout.flush()
         try:
-            subprocess.check_call(run_command)
+            proc = subprocess.run(run_command, check=True, stdout=PIPE, stderr=PIPE)
+            return proc.stdout
```
Instead of returning, can you simply add a stdouts attribute to the parameters class at the top of this file. You can then simply append to self.parameters.stdouts, and then there is no reason to have to pass it in to analse function.

pgrete replied:
Great point! That make this much cleaner.

JoshuaSBrown commented on tst/regression/utils/test_case.py:
```diff
@@ -207,20 +208,21 @@ def Run(self):
                               .format(err.returncode, ' '.join(err.cmd)))
 
-    def Analyse(self):
+    def Analyse(self, stdouts):
```
Follow instructions above and then remove stdouts argument.

JoshuaSBrown commented on tst/regression/utils/test_case.py:
```diff
@@ -207,20 +208,21 @@ def Run(self):
         print("*****************************************************************")
         sys.stdout.flush()
-        test_pass = self.test_case.Analyse(self.parameters)
+        test_pass = self.test_case.Analyse(self.parameters, stdouts)
```
Follow instructions above and remove stdouts argument, it should be available in self.parameters.

JoshuaSBrown commented on tst/regression/CMakeLists.txt:
```diff
@@ -110,3 +117,9 @@ foreach( index RANGE ${end_index})
   setup_test(${TEST_DIR} ${TEST_ARG})
   setup_test_mpi(${TEST_PROC} ${TEST_DIR} ${TEST_ARG})
 endforeach()
+
+# manually add label to performance regression test (to selectively enable/disable)
+set_tests_properties(regression_test:performance PROPERTIES LABELS "regression;mpi-no;perf-reg" )
+if( MPI_FOUND )
+  set_tests_properties(regression_mpi_test:performance PROPERTIES LABELS "regression;mpi-yes;perf-reg" )
+endif()
```
This is fine as it is, but it points to an underlying problem with how I implemented the cmake script in this case. The above foreach loop and functions would ideally be flexible enough to allow for extra labels and other customization without having to introduce if branches like this, which makes the script less readable. I will create an issue to come up with a better design.

pgrete commented:
Thanks for the review. I addressed the changes (parsing of `stdout`, directory naming, added doc) and updated to `master`.
Let me know what you think.

pgrete commented:
ping @jdolence (or someone else from the LANL physics team) 
Any input on this PR?

Yurlungur commented:
I'll take a look this afternoon.

Yurlungur commented on tst/regression/CMakeLists.txt:
```diff
@@ -53,6 +53,13 @@ list(APPEND TEST_ARGS "--driver ${CMAKE_BINARY_DIR}/example/advection/advection-
+# Performance regression test
+list(APPEND TEST_DIRS advection_performance)
+list(APPEND TEST_PROCS "1")
+list(APPEND TEST_ARGS "--driver ${CMAKE_BINARY_DIR}/example/advection/advection-example \
+--driver_input ${CMAKE_CURRENT_SOURCE_DIR}/test_suites/advection_performance/parthinput.advection_performance \
+--num_steps 5")
```
This does not seem consistent with the documentation. It looks like we have to manually add tests to `cmake`?

pgrete replied:
Yes. Good point. I'll update the doc.

pgrete commented:
Updated the docs, test pass, got two approval, pulling the trigger.
 @JoshuaSBrown  feel free to put the docker file to the other repo, too, pending the result of the discussion in #200
